Hit testing on element with 180deg flip inside preserve-3d not finding correct element.

mattwoodrow opened the pull request:
#### 1aa517d47471fdc37ac5b6e2f3809de704ea8ac5
<pre>
Hit testing on element with 180deg flip inside preserve-3d not finding correct element.
<a href="https://bugs.webkit.org/show_bug.cgi?id=258565">https://bugs.webkit.org/show_bug.cgi?id=258565</a>
&lt;rdar://111393557&gt;

Reviewed by Simon Fraser.

Floating point precision (using double) meant that sin(180deg) wasn&apos;t returning exactly 0.

When combining multiple transforms together using preserve-3d, this meant that layers that are intended to be coplanar
ended up at slightly different depths.

This adds rounding for tiny values to make them exactly zero, so that the layers end up coplanar again, and we hit
test in DOM order for those.

* LayoutTests/transforms/3d/hit-testing/hit-preserves-3d-3-expected.txt: Added.
* LayoutTests/transforms/3d/hit-testing/hit-preserves-3d-3.html: Added.
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
(WebCore::roundEpsilonToZero):
(WebCore::TransformationMatrix::rotate3d):
(WebCore::TransformationMatrix::rotate):

Canonical link: <a href="https://commits.webkit.org/266237@main">https://commits.webkit.org/266237@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/581ad8df260cd22d78deb468da70ce4214bd02fd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13318 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13631 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13964 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15055 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12679 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13397 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16139 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13658 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15360 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13485 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14141 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11258 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15684 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11429 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12013 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19067 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12504 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12180 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15393 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12685 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10544 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11956 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11877 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3243 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16278 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12527 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
```diff
@@ -907,6 +908,13 @@ TransformationMatrix& TransformationMatrix::scale3d(double sx, double sy, double
 
+static double roundEpsilonToZero(double val)
+{
+    if (-FLT_EPSILON < val && val < FLT_EPSILON)
```
For `double` I believe we want [`DBL_EPSILON`](https://en.cppreference.com/w/c/types/limits#:~:text=FLT_EPSILON-,DBL_EPSILON,-LDBL_EPSILON)?

smfr commented on Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
```diff
@@ -907,6 +908,13 @@ TransformationMatrix& TransformationMatrix::scale3d(double sx, double sy, double
     return *this;
 }
 
+static double roundEpsilonToZero(double val)
```
Should this use `areEssentiallyEqual()` in MathExtras.h?

mattwoodrow replied:
Probably, yes :) Will change that!

mattwoodrow replied:
Ah nope, it doesn't consider them close enough to be equal :(

smfr commented on Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
```diff
@@ -923,8 +931,8 @@ TransformationMatrix& TransformationMatrix::rotate3d(double x, double y, double
 
-    double sinTheta = sin(angle);
-    double cosTheta = cos(angle);
+    double sinTheta = roundEpsilonToZero(sin(angle));
```
Not sure if this should be down here in TransformationMatrix or be the responsibility of callers who care about being close to zero?

mattwoodrow replied:
I think it should be here, or possibly at some point in an inner helper that does degrees -> sin/cos conversion.

This TransformationMatrix function takes an angle in degrees, and wants the rotation matrix for it. The bug is introduced by going via radians before computing the matrix vectors, and losing precision slightly. My take is that's a bug in the implementation detail and no caller would want the other behaviour.

mattwoodrow committed:
Hit testing on element with 180deg flip inside preserve-3d not finding correct element.
https://bugs.webkit.org/show_bug.cgi?id=258565
<rdar://111393557>

Reviewed by Simon Fraser.

Floating point precision (using double) meant that sin(180deg) wasn't returning exactly 0.

When combining multiple transforms together using preserve-3d, this meant that layers that are intended to be coplanar
ended up at slightly different depths.

This adds rounding for tiny values to make them exactly zero, so that the layers end up coplanar again, and we hit
test in DOM order for those.

* LayoutTests/transforms/3d/hit-testing/hit-preserves-3d-3-expected.txt: Added.
* LayoutTests/transforms/3d/hit-testing/hit-preserves-3d-3.html: Added.
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
(WebCore::roundEpsilonToZero):
(WebCore::TransformationMatrix::rotate3d):
(WebCore::TransformationMatrix::rotate):

Canonical link: https://commits.webkit.org/266237@main
